### ISSUE TITLE
在ppo.py中计算values时，是何时把merge进chatglm2-6b的sft给分离出来的？

### ISSUE THREAD

**Issue by lhzwangwei** [hiyouga/ChatGLM-Efficient-Tuning#313]
训练rm时不是把chatglm2-6b加上rm的lora以及valuehead来训练的么，在进行ppo时，在计算reward之间，不是应该把sft从模型上unmerge出来么？
` replace_model(unwrapped_model, target="reward")
                _, _, values = self.model(**self.prepare_model_inputs(queries, responses))
                rewards = [reward for reward in values[-1].to(torch.float32)] # use float32 type
                replace_model(unwrapped_model, target="default") # make sure the model is default at the end`
实在这个replace_model中么？
`    def set_adapter(self, adapter_name):
        for module in self.model.modules():
            if isinstance(module, LoraLayer):
                if module.merged:
                    warnings.warn("Adapter cannot be set when the model is merged. Unmerging the model first.")
                    module.unmerge()
                module.active_adapter = adapter_name`

但是这个module.unmerge()没有执行。

**Comment by lhzwangwei** [comment]
合并sft时，是直接用merge_and_unload把原来的继承了Loralayer的Linear给直接替换成torch.nn.linear了，那这样就不能在unmerge了。求大佬解惑。

**Comment by lhzwangwei** [comment]
顶一顶

**Comment by huhubrave** [comment]
顶一顶。我也有类似疑惑。   希望有谁能解答疑惑。感谢。
reward model 在训练时，是基于base model的。而在PPO阶段，reward model的推理 是基于SFT模型的，因为在一开始base model和SFT的lora权重就merge_and_unload() @hiyouga

**Comment by hiyouga** [comment]
reward model 应该基于 sft 模型训练才对

**Comment by lhzwangwei** [comment]
但是原本的关于rm训练的代码中，只是在base model 上加载了一个lora ，并没有先把第一步sft的lora给merge()进base model

**Comment by lhzwangwei** [comment]
是最新更新的代码里这点已经修改了么？还没下载最新的

**Comment by hiyouga** [comment]
把训练时候的命令行参数修改一下就行

**Comment by lhzwangwei** [comment]
谢谢大佬回复。就是在训练rm时，先把第一个sft的loramerge()进来，再进行训练对么？

**Comment by hiyouga** [comment]
是的
